src/feature/mod.rs

```diff
@@ -525,46 +635,41 @@ pub fn restore(trans: &postgres::transaction::Transaction, schema: &Option<valic
         Err(_) => { return Err(import_error(&feat, "Failed to stringify properties", None)) }
     };
 
-    //Get the previous version of a given feature
+    //Get the previous version, action, and all deltas of a given feature
     match trans.query("
+    SELECT
+        ARRAY_AGG(delta ORDER BY delta) as delta_ids,
+        MAX(version) as max_version,
+        (ARRAY_AGG(action ORDER BY version DESC))[1] as action
+    FROM (
         SELECT
-            ARRAY_AGG(id ORDER BY id) AS delta_ids,
-            MAX(feat->>'version')::BIGINT + 1 AS max_version
-        FROM (
-            SELECT
-                deltas.id,
-                JSON_Array_Elements((deltas.features -> 'features')::JSON) AS feat
-            FROM
-                deltas
-            WHERE
-                affected @> ARRAY[$1]::BIGINT[]
-            ORDER BY id DESC
-        ) f
+            delta,
+            version,
+            action
+        FROM
+            geo_history
         WHERE
-            (feat->>'id')::BIGINT = $1
-        GROUP BY feat->>'id'
+            id=$1
+    ) t
     ", &[&id]) {
-        Ok(history) => {
+        Ok(res) => {
+            // Agg function returns row of NULLs if inner query doesn't return results
+            let affected: Option<Vec<i64>> = res.get(0).get(0);
+            let affected = match affected {
+                None => return Err(import_error(&feat, "Feature Not Found", None)),
+                Some(deltas) => deltas
+            };
 
-            if history.len() != 1 {
-                return Err(import_error(&feat, "Feature Not Found", None));
+            let prev_version: i64 = res.get(0).get(1);
+            // previous version of a feature must match version passed in request
+            if prev_version != version {
+                return Err(import_error(&feat, "Restore Version Mismatch", None));
             }
 
-            //Version will be None if the feature was created but has never been modified since the
-            //original create does not need a version
-            let prev_version: Option<i64> = history.get(0).get(1);
-            match prev_version {
-                None => {
-                    return Err(import_error(&feat, "Feature Not In Deleted State", None));
-                },
-                Some(prev_version) => {
-                    if prev_version != version {
-                        return Err(import_error(&feat, "Restore Version Mismatch", None));
-                    }
-                }
-            };
-
-            let affected: Vec<i64> = history.get(0).get(0);
+            let prev_action: String = res.get(0).get(2);
+            if prev_action != String::from("delete") {
```

Comment:
This now checks that the last action was a `delete` rather than just checking that the feature's most recent modification was a `create`